sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala

```diff
@@ -35,184 +34,289 @@ import org.apache.spark.util.Utils
+    case st: StructType =>
+      val fieldConverters = st.map(_.dataType).map(makeConverter)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_OBJECT => convertObject(parser, st, fieldConverters)
+      }
 
-      case (START_ARRAY, ArrayType(st, _)) =>
-        convertArray(factory, parser, st)
+    case at: ArrayType =>
+      val elementConverter = makeConverter(at.elementType)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_ARRAY => convertArray(parser, elementConverter)
+      }
 
-      case (START_OBJECT, MapType(StringType, kt, _)) =>
-        convertMap(factory, parser, kt)
+    case mt: MapType =>
+      val valueConverter = makeConverter(mt.valueType)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_OBJECT => convertMap(parser, valueConverter)
+      }
 
-      case (_, udt: UserDefinedType[_]) =>
-        convertField(factory, parser, udt.sqlType)
+    case udt: UserDefinedType[_] =>
+      makeConverter(udt.sqlType)
 
-      case (token, dataType) =>
-        // We cannot parse this token based on the given data type. So, we throw a
-        // SparkSQLJsonProcessingException and this exception will be caught by
-        // parseJson method.
-        throw new SparkSQLJsonProcessingException(
-          s"Failed to parse a value for data type $dataType (current token: $token).")
+    case _ =>
+      (parser: JsonParser) =>
+        parseJsonToken(parser, dataType) {
+          case token =>
+            throw new SparkSQLJsonProcessingException(
+              s"Failed to parse a value for data type $dataType (current token: $token).")
+        }
+  }
+
+  /**
+   * This handles nulls ahead before trying to check the tokens, and applies the conversion
+   * function and then checks failed the conversion afterward if it `f` fails to convert the value.
+   *
+   * In more details, it checks `FIELD_NAME` if exists and then skip. If this is called after
+   * `START_OBJECT`, then, the next token can be `FIELD_NAME`. Since the names are kept in
+   * `JacksonParser.convertObject`, this `FIELD_NAME` token can be skipped as below. When this
+   * is called after `START_ARRAY`, the tokens become ones about values until `END_ARRAY`.
+   * In this case, we don't have to skip.
+   *
+   * We check if the current token is null or not after that. Then, we apply `f` to convert
+   * the value and then we check failed conversion afterward if it `f` fails to convert the value.
+   */
+  private def parseJsonToken(
+      parser: JsonParser,
+      dataType: DataType)(f: PartialFunction[JsonToken, Any]): Any = {
+    parser.getCurrentToken match {
+      case FIELD_NAME =>
+        parser.nextToken()
+        parseJsonToken(parser, dataType)(f)
+
+      case null | VALUE_NULL => null
+
+      case other => f.orElse {
+        // We should specify the type of this `PartialFunction`. Otherwise this will
```

Comment:
instead of adding comments to explain this hack, I'd like to avoid it. We can still use the method `failedConversion`, and use `f.applyOrElse(failedConversion(...))` here